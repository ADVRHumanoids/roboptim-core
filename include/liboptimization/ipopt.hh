// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Declaration of the IpoptSolver class.
 */

#ifndef OPTIMIZATION_IPOPT_HH
# define OPTIMIZATION_IPOPT_HH
# include <coin/IpSmartPtr.hpp>
# include <liboptimization/solver.hh>
# include <liboptimization/twice-derivable-function.hh>

/// Ipopt classes.
namespace Ipopt
{
  class TNLP;
  class IpoptApplication;
}; // end of namespace Ipopt


namespace optimization
{
  namespace detail
  {
    class MyTNLP;
  };

  /// Ipopt based solver.
  class IpoptSolver : public Solver<TwiceDerivableFunction,
                                    const TwiceDerivableFunction*>
  {
  public:
    friend class detail::MyTNLP;
    typedef Solver<TwiceDerivableFunction, const TwiceDerivableFunction*> parent_t;

    /// Constructor.
    explicit IpoptSolver (const problem_t&) throw ();
    /// Destructor.
    virtual ~IpoptSolver () throw ();

    /// Solve the problem.
    virtual void solve () throw ();

    /// Get Ipopt Application object for Ipopt specific tuning.
    virtual Ipopt::SmartPtr<Ipopt::IpoptApplication> getApplication ()
      throw ();
  private:
    /// Smart pointer to the Ipopt non linear problem description.
    Ipopt::SmartPtr<Ipopt::TNLP> nlp_;
    /// Smtart pointer to the Ipopt application instance.
    Ipopt::SmartPtr<Ipopt::IpoptApplication> app_;
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_IPOPT_HH
