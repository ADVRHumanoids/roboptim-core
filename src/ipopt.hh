// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file src/ipopt.hh
 *
 * \brief Declaration of the IpoptSolver class.
 */

#ifndef OPTIMIZATION_IPOPT_HH
# define OPTIMIZATION_IPOPT_HH
# include <IpSmartPtr.hpp>
# include <solver.hh>

/// Ipopt classes.
namespace Ipopt
{
  class TNLP;
  class IpoptApplication;
}; // end of namespace Ipopt


namespace optimization
{
  namespace detail
  {
    class MyTNLP;
  };

  /// Ipopt based solver.
  class IpoptSolver : public Solver
  {
  public:
    friend class detail::MyTNLP;

    /// Constructor.
    explicit IpoptSolver (const Problem&) throw ();
    /// Destructor.
    virtual ~IpoptSolver () throw ();

    /// Returns the minimum and solve the problem using Ipopt if necessary.
    virtual result_t getMinimum () throw ();

    /// Get Ipopt Application object for Ipopt specific tuning.
    virtual Ipopt::SmartPtr<Ipopt::IpoptApplication> getApplication ()
      throw ();
  private:
    /// Smart pointer to the Ipopt non linear problem description.
    Ipopt::SmartPtr<Ipopt::TNLP> nlp_;
    /// Smtart pointer to the Ipopt application instance.
    Ipopt::SmartPtr<Ipopt::IpoptApplication> app_;
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_IPOPT_HH
