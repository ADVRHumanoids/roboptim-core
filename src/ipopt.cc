// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file ipopt.cc
 *
 * \brief Implementation of the Ipopt class.
 */

#include <cassert>

#include <IpIpoptApplication.hpp>
#include <IpTNLP.hpp>

#include "ipopt.hh"

namespace optimization
{
  using namespace Ipopt;

  namespace detail
  {
    struct MyTNLP : public TNLP
    {
      MyTNLP (IpoptSolver& solver)
        : solver_ (solver)
      {
      }

      virtual bool
      get_nlp_info (Index& n, Index& m, Index& nnz_jac_g,
                    Index& nnz_h_lag, TNLP::IndexStyleEnum& index_style)
      {
        n = solver_.getArity ();
        m = 0;
        nnz_jac_g = 0;
        nnz_h_lag = 0;
        index_style = TNLP::C_STYLE;
        return true;
      }
      virtual bool
      get_bounds_info (Index n, Number* x_l, Number* x_u,
                       Index m, Number* g_l, Number* g_u)
      {
        return true;
      }

      virtual bool
      get_starting_point(Index n, bool init_x, Number* x,
                         bool init_z, Number* z_L, Number* z_U,
                         Index m, bool init_lambda,
                         Number* lambda)
      {
        return true;
      }

      virtual bool
      eval_f (Index n, const Number* x, bool new_x, Number& obj_value)
      {
        IpoptSolver::array_t x_ (n);

        // FIXME: check if this is safe.
        memcpy (&x_[0], x, n * sizeof (Number));

        obj_value = solver_.getFunction () (x_);
        return true;
      }

      virtual bool
      eval_grad_f(Index n, const Number* x, bool new_x, Number* grad_f)
      {
        return true;
      }

      virtual bool
      eval_g(Index n, const Number* x, bool new_x,
                        Index m, Number* g)
      {
        return true;
      }

      virtual bool
      eval_jac_g(Index n, const Number* x, bool new_x,
                 Index m, Index nele_jac, Index* iRow,
                 Index *jCol, Number* values)
      {
        return true;
      }

      virtual void
      finalize_solution(SolverReturn status,
                        Index n, const Number* x, const Number* z_L, const Number* z_U,
                        Index m, const Number* g, const Number* lambda,
                        Number obj_value,
                        const IpoptData* ip_data,
                        IpoptCalculatedQuantities* ip_cq)
      {
        if (status != SUCCESS)
          {
            solver_.result_ = SolverError ();
            return;
          }

        solver_.result_ = obj_value;
      }

      IpoptSolver& solver_;
    };
  }; // end of namespace detail

  using namespace detail;

  IpoptSolver::IpoptSolver (const function_t& fct, size_type n) throw ()
    : Solver (fct, n),
      nlp_ (new MyTNLP (*this)),
      app_ (new IpoptApplication ())
  {
  }

  IpoptSolver::~IpoptSolver () throw ()
  {
  }

  IpoptSolver::result_t
  IpoptSolver::getMinimum () throw ()
  {
    if (result_.which () != SOLVER_NO_SOLUTION)
      return result_;

    ApplicationReturnStatus status = app_->Initialize ();
    if (status != Solve_Succeeded)
      {
        result_ = SolverError ();
        return result_;
      }
    app_->OptimizeTNLP (nlp_);
    return result_;
  }

} // end of namespace optimization
