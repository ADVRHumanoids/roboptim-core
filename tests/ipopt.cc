// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>
#include <boost/lambda/lambda.hpp>
#include <boost/numeric/ublas/io.hpp>
#include <boost/variant/get.hpp>

#include <coin/IpIpoptApplication.hpp>

#include <liboptimization/ipopt.hh>

#include "common.hh"
#include "hs071.hh"

int run_test ()
{
  F f;
  G0 g0;
  G1 g1;

  IpoptSolver::problem_t pb (f);
  pb.constraints ().push_back (&g0);
  pb.constraints ().push_back (&g1);

  // Set the starting point.
  Function::vector_t start (f.n);
  start[0] = 1., start[1] = 5., start[2] = 5., start[3] = 1.;
  pb.startingPoint () = start;

  // Initialize solver
  IpoptSolver solver (pb);

  // Compute the minimum and retrieve the result.
  IpoptSolver::result_t res = solver.minimum ();

  // Display solver information.
  std::cout << solver << std::endl;

  // Check if the minimization has succeed.
  if (res.which () != IpoptSolver::SOLVER_VALUE)
    {
      std::cout << "A solution should have been found. Failing..."
                << std::endl;
      return 1;
    }

  // Get the result.
  IpoptSolver::vector_t& result = boost::get<IpoptSolver::vector_t> (res);

  // Display the result.
  std::cout << "A solution has been found: " << std::endl;
  std::cout << result << std::endl;
  std::cout << "f(*x) = " << solver.problem ().function () (result) << std::endl;
  return 0;
}


GENERATE_TEST ()
