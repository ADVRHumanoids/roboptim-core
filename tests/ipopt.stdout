Twice derivable function
Number of constraints: 2
Twice derivable function
Twice derivable function
A solution has been found: 
[4](1,4.743,3.82115,1.37941)
f(*x) = 14.5723
